src/services.py

```diff
@@ -0,0 +1,1033 @@
+# Copyright 2024 Canonical Ltd.
```

Comment:
[Suggestion] Having a file with thousand lines of code is not a good practice especially if it contains multiple classes. I'd suggest you to have python package called `services`. Under this package, create modules `abstract_service.py`, `upgrade_service.py`, etc. for each class. This will increase both maintainability and readability.

Reply:
Now that I think about it, I agree. However, the charm will be broken up into other pieces relatively soon and I don't want to spend the time on this right now. I will leave a TODO and shelve this for now.